core/subscription/subscription_engine.hpp

```diff
@@ -0,0 +1,99 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_SUBSCRIPTION_ENGINE_HPP
+#define KAGOME_SUBSCRIPTION_ENGINE_HPP
+
+#include <list>
+#include <memory>
+#include <shared_mutex>
+#include <unordered_map>
+
+namespace kagome::subscription {
+  template <typename Key, typename Type, typename... Arguments>
+  class Subscriber;
+
+  template <typename Key, typename Type, typename... Arguments>
+  class SubscriptionEngine final
+      : public std::enable_shared_from_this<
+            SubscriptionEngine<Key, Type, Arguments...>> {
+   public:
+    using KeyType = Key;
+    using ValueType = Type;
+    using SubscriberType = Subscriber<KeyType, ValueType, Arguments...>;
+    using SubscriberPtr = std::shared_ptr<SubscriberType>;
+    using SubscriberWPtr = std::weak_ptr<SubscriberType>;
+
+    /// List preferable here because of remains the whole containers iterators
+    /// alive after remove from the middle of the container
```

Comment:
```suggestion
    /// List is preferable here because this container iterators remain
    /// alive after removal from the middle of the container
```

Reply:
ok